### PR TITLE
use grocksdb to support latest version of rocksdb

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #43 #44 
also add API `NewRocksDBWithRawDB` to support read-only open mode and customize options.

The impact on existing user is it requires to build with rocksdb `v7.10.2` now, otherwise, it behaves the same as before.

### CODE REVIEW

**Comment by thanethomson** [comment]
@yihuang, could you please provide a more detailed explanation (ideally in an issue) as to why this change is necessary?

Also, it appears as though the tests in CI are failing.

**Comment by yihuang** [comment]
> @yihuang, could you please provide a more detailed explanation (ideally in an issue) as to why this change is necessary?
> 
> Also, it appears as though the tests in CI are failing.

done, added issue #43 and #44 which are fixed by this PR at once.

**Comment by yihuang** [comment]
> Thanks for this proposal @yihuang !
> 
> Copying @faddat, do you see any concerns in replacing github.com/tecbot/gorocksdb with github.com/linxGnu/grocksdb ?

we already used on production, so far so good, the only thing different is the building dependency change, no side-effect observed in runtime, other than the changes caused by our tuning of rocksdb options.

**Comment by adizere** [comment]
Re: the CI failure for the "Docker testing image" workflow --

Not sure what the fix is, but I managed to repro the same problem in a separate PR (#45). Is it possible that [the DockerHub secrets](https://github.com/cometbft/cometbft-db/actions/runs/4361242056/workflow#L58-L60) are not propagated into forks (such [as this](https://github.com/yihuang/cometbft-db)) of this repo and that is causing the failure?

**Comment by thanethomson** [comment]
@yihuang please address the failing test. This failure is independent of the Docker build failure (we're investigating what's causing that).

**Comment by yihuang** [comment]
> @yihuang please address the failing test. This failure is independent of the Docker build failure (we're investigating what's causing that).

I think it's just due to incompatible rocksdb dependency.
```
/go/pkg/mod/github.com/linx!gnu/grocksdb@v1.7.15/cache.go:86:2: could not determine kind of name for C.rocksdb_lru_cache_options_set_num_shard_bits
```

**Review by robert-zaremba — APPROVED** [review]
tecbot/gorocksdb is outdated, and missing a maintainer.

`linxGnu/grocksdb` is definitely a way to go.

* pending test fixes

**Comment by facundomedica** [comment]
> > @yihuang please address the failing test. This failure is independent of the Docker build failure (we're investigating what's causing that).
> 
> I think it's just due to incompatible rocksdb dependency, which can be fixed by updating the image with newer rocksdb library.
> 
> ```
> /go/pkg/mod/github.com/linx!gnu/grocksdb@v1.7.15/cache.go:86:2: could not determine kind of name for C.rocksdb_lru_cache_options_set_num_shard_bits
> ```

Looks like that was added on `7.5.3` any ideas why it's failing? Right now building Cosmos SDK with Rocks DB is not possible without using a fork + some replaces :(

File: tools/Dockerfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 
 FROM build AS install
 ARG LEVELDB=1.20
-ARG ROCKSDB=6.24.2
+ARG ROCKSDB=7.9.2
```
Comment (alexanderbez) [inline tools/Dockerfile:20]: Latest version is now v8 BTW

**Review by alexanderbez — APPROVED** [review]
utACK. Let's get this merged and released ASAP :)

File: tools/Dockerfile (hunk 1 above)
Comment (alexanderbez) [reply tools/Dockerfile:20]: https://github.com/facebook/rocksdb/releases/tag/v8.0.0

**Comment by yihuang** [comment]
> > > @yihuang please address the failing test. This failure is independent of the Docker build failure (we're investigating what's causing that).
> 
> > 
> 
> > I think it's just due to incompatible rocksdb dependency, which can be fixed by updating the image with newer rocksdb library.
> 
> > 
> 
> > ```
> 
> > /go/pkg/mod/github.com/linx!gnu/grocksdb@v1.7.15/cache.go:86:2: could not determine kind of name for C.rocksdb_lru_cache_options_set_num_shard_bits
> 
> > ```
> 
> 
> 
> Looks like that was added on `7.5.3` any ideas why it's failing? Right now building Cosmos SDK with Rocks DB is not possible without using a fork + some replaces :(

I guess become the CI image not updated?

File: tools/Dockerfile (hunk 1 above)
Comment (yihuang) [reply tools/Dockerfile:20]: https://github.com/linxGnu/grocksdb/pull/111
Need a few tweaks to support v8.0

Comment (alexanderbez) [reply tools/Dockerfile:20]: you rock!

**Comment by yihuang** [comment]
> @yihuang we updated the CI to now use a Docker image built from the _current_ branch (as opposed to one previously built from Docker Hub) and it still seems as though the test is failing.

CI passed now.

**Review by lasarojc — APPROVED** [review]

**Comment by thanethomson** [comment]
How do folks using rocksdb with the current cometbft-db release upgrade?

File: Makefile [hunk 2]
```diff
@@ -43,7 +43,7 @@ test-badgerdb:
 
 test-all:
 	@echo "--> Running go test"
-	@go test $(PACKAGES) -tags cleveldb,boltdb,rocksdb,badgerdb -v
+	@go test $(PACKAGES) -tags cleveldb,boltdb,rocksdb,grocksdb_clean_link,badgerdb -v
```
Comment (thanethomson) [inline Makefile:46]: What does this `grocksdb_clean_link` tag do?

Comment (yihuang) [reply Makefile:46]: It's to control the link flags, the default one is suitable for static linking which specifies all indirect dependencies explicitly, the clean version is suitable for dynamic linking case which only specify "-lrocksdb"

**Comment by robert-zaremba** [comment]
Does this fix the rocksdb usage with Cosmos SDK 0.46 or 0.47?

**Comment by lasarojc** [comment]
> How do folks using rocksdb with the current cometbft-db release upgrade?

Apparently simply upgrading should be fine. According to 
https://github.com/facebook/rocksdb/wiki/RocksDB-Compatibility-Between-Different-Releases

> Backward compatible: newer version of RocksDB should be able to open DBs generated by all previous releases for normal configuration.

**Comment by yihuang** [comment]
> How do folks using rocksdb with the current cometbft-db release upgrade?

In our experience, upgrading rocksdb itself is totally backward compatible.

**Review by thanethomson — COMMENTED** [review]
> In our experience, upgrading rocksdb library is totally backward compatible.

Excellent. If that's the case, please add a changelog entry under the "Dependencies" section (as per https://github.com/cometbft/cometbft/blob/main/CONTRIBUTING.md#changelog) and then we can merge and ship this in a non-breaking release.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # CHANGELOG
 
+## UNRELEASED
+
+### Dependencies
+
+- Switch rocksdb binding from gorocksdb to grocksdb, update librocksdb dependency to `v7.10.2`, [#42](https://github.com/cometbft/cometbft-db/pull/42).
+
```
Comment (thanethomson) [inline CHANGELOG.md:8]: Please see the contributing guidelines for how to add a changelog entry: https://github.com/cometbft/cometbft/blob/main/CONTRIBUTING.md#changelog

It needs to exist as a separate file in the `.changelog/unreleased/dependencies` directory.

Comment (yihuang) [reply CHANGELOG.md]: done

**Pull request merged**